src/query-builder/InsertQueryBuilder.ts

```diff
@@ -427,7 +427,7 @@ export class InsertQueryBuilder<Entity> extends QueryBuilder<Entity> {
         // if column metadatas are given then apply all necessary operations with values
         if (columns.length > 0) {
             let expression = "";
-            let parametersCount = Object.keys(this.expressionMap.nativeParameters).length;
```

Comment:
I don't remember why this line of code was needed, but I don't think it was incidentally placed here. There is definitely a mess around parameters in QueryBuilder, and this area of code needs to be refactored. I'm not ready to break any exist functionality in order to fix `printSql` method. We'll need to figure out why this line of code was needed before we can make a decision about this change.

Reply:
Thanks for the feedback! Let me take another look at this and see if I can either give us more confidence or find an alternative implementation.

Reply:
This is here for cases when other parts of the query builder have already added parameters to `expressionMap.nativeParameters`. In certain drivers it may be possible that some are already added, even with an `INSERT` query so we need to maintain the correct index/order. `this.connection.driver.createParameter()` creates the `$X` string based on whatever number is given to it.

I've mostly resolved the ordering problem in #7282, but it still remains that `nativeParameters` is shared between multiple `printSql()`/`getSql()` calls, leading to the issue described. We need to move it to a temporary storage that isn't `expressionMap`, as the parameters are calculated every time the query is compiled.